controllers/controller_shared.go

```diff
@@ -366,3 +367,17 @@ func getReferencedValue(ctx context.Context, cl client.Client, cr metav1.ObjectM
 		}
 	}
 }
+
+// Add finalizer through a MergePatch
+// Avoids updating the entire object and only changes the finalizers
+func addFinalizer(ctx context.Context, cl client.Client, cr client.Object) error {
+	patch := []byte(`{"metadata":{"finalizers":["` + grafanaFinalizer + `"]}}`)
+	return cl.Patch(ctx, cr, client.RawPatch(types.MergePatchType, patch))
+}
+
+// Remove finalizer through a MergePatch
+// Avoids updating the entire object and only changes the finalizers
+func removeFinalizer(ctx context.Context, cl client.Client, cr client.Object) error {
+	patch := []byte(`{"metadata":{"finalizers":null}}`)
```

Comment:
Although it's nice that you can fix the drift with these changes, I would disagree with the way how patches are formed. - Kubernetes objects may have multiple finalizers, whereas, in this PR, there is an assumption that there will always be a single finalizer owned by the grafana-operator. - Some users might have their own finalizers (think of custom operators).

Since `controllerutil` has two functions that safely add (`AddFinalizer`) / remove (`RemoveFinalizer `) a particular finalizer, you can, for instance, have them modify finalizers and then prepare a patch by copying value from `metadata.finalizers`.

Reply:
Sure, I'll try and whip up a solution that makes sense that safely patches the finalizers.

Reply:
I ended up almost copying the `controllerUtil.RemoveFinalizer` implementation for removing a finalizer as it worked quite well.

Reply:
I'm not sure if we should copy code from anywhere. Do you see any benefits compared to direct calls to their library? - Like it was mentioned in my previous comment, I thought of rather calling `controllerutil.AddFinalizer` / `controllerutil.RemoveFinalizer` and then copying `finalizers` field from the result. - It's pretty straightforward, we don't have to maintain code of the underlying functions, and it doesn't create licensing complexity (I think you have to mention authors when copying Apache2 code).

Reply:
I partially misunderstood your previous message.
This is iteration aligns with your suggestions!